src/main/java/org/springframework/integration/kafka/inbound/KafkaMessageDrivenChannelAdapter.java

```diff
@@ -192,36 +202,29 @@ public void doOnMessage(Object key, Object payload, KafkaMessageMetadata metadat
 	private Message<Object> toMessage(Object key, Object payload, KafkaMessageMetadata metadata,
 			Acknowledgment acknowledgment) {
 
-		final Map<String, Object> headers = new HashMap<String, Object>();
+		final MessageHeaderAccessor headerAccessor = new MessageHeaderAccessor();
 
-		headers.put(KafkaHeaders.MESSAGE_KEY, key);
-		headers.put(KafkaHeaders.TOPIC, metadata.getPartition().getTopic());
-		headers.put(KafkaHeaders.PARTITION_ID, metadata.getPartition().getId());
-		headers.put(KafkaHeaders.OFFSET, metadata.getOffset());
-		headers.put(KafkaHeaders.NEXT_OFFSET, metadata.getNextOffset());
+		headerAccessor.setHeader(KafkaHeaders.MESSAGE_KEY, key);
+		headerAccessor.setHeader(KafkaHeaders.TOPIC, metadata.getPartition().getTopic());
+		headerAccessor.setHeader(KafkaHeaders.PARTITION_ID, metadata.getPartition().getId());
+		headerAccessor.setHeader(KafkaHeaders.OFFSET, metadata.getOffset());
+		headerAccessor.setHeader(KafkaHeaders.NEXT_OFFSET, metadata.getNextOffset());
 
 		// pre-set the message id header if set to not generate
-		if (!this.generateMessageId) {
-			headers.put(MessageHeaders.ID, MessageHeaders.ID_VALUE_NONE);
-		}
-
-		// pre-set the timestamp header if set to not generate
-		if (!this.generateTimestamp) {
-			headers.put(MessageHeaders.TIMESTAMP, -1L);
-		}
+		 headerAccessor.setLeaveMutable(!(this.generateMessageId || this.generateTimestamp));
 
 		if (!this.autoCommitOffset) {
-			headers.put(KafkaHeaders.ACKNOWLEDGMENT, acknowledgment);
+			headerAccessor.setHeader(KafkaHeaders.ACKNOWLEDGMENT, acknowledgment);
 		}
 
 		if (this.useMessageBuilderFactory) {
 			return getMessageBuilderFactory()
 					.withPayload(payload)
-					.copyHeaders(headers)
+					.copyHeaders(headerAccessor.toMessageHeaders())
 					.build();
 		}
 		else {
-			return new KafkaMessage(payload, headers);
```

Comment:
M-m-m. @markpollack , the pattern looks like:
```
MessageHeaderAccessor headerAccessor = new MessageHeaderAccessor();
```
Which does:
```
public MessageHeaderAccessor() {
	this.headers = new MutableMessageHeaders();
}
```
Than we add desired headers like this:
```
headerAccessor.setHeader(KafkaHeaders.TOPIC, metadata.getPartition().getTopic());
```
And before message creation:
```
headerAccessor.setLeaveMutable(!(this.generateMessageId || this.generateTimestamp));
```
Which impacts to this:
```
return MessageBuilder.createMessage(payload, headerAccessor.getMessageHeaders());
```
where we have:
```
return new GenericMessage<T>(payload, messageHeaders);
```
just with this:
```
public GenericMessage(T payload, MessageHeaders headers) {
	Assert.notNull(payload, "Payload must not be null");
	Assert.notNull(headers, "MessageHeaders must not be null");
	this.payload = payload;
	this.headers = headers;
}
```

Have I missed anything?

If you do a performance testing from XD, maybe it ends up with this branch of logic:
```
if (this.useMessageBuilderFactory) {
	return getMessageBuilderFactory()
			.withPayload(payload)
			.copyHeaders(headerAccessor.toMessageHeaders())
			.build();
}
```
Where we really do the copy.
Need to know Kafka source config on the matter.

From other side we really might broke some logic, when the best part had been by default before.